Implement the docs' snippet framework transformer.

jansiegel opened the pull request:
### Context
This PR implements a "snippet transformer", which is supposed to be a part of the documentation building process.

Its main job is to take the code blocks defined in the documentation's `.md` files and return a snippet in the desired framework's template.

It does so by:
### 1. Introducing a new Vuepress container, `::: snippet`:
Any code block inside a ` ```js` or ` ```javascript` block wrapped in the `::: snippet` container will be picked up and transformed by the transformer.

The final snippet will **not** include the imports and the app structure, as that code is not meant to be executed directly. 

For example:
   <pre><code>::: snippet
```js
  const data = [['test', 'data']]; 
  const hot = new Handsontable(container, {
      data,
      nestedRows: true,
      afterRender: onAfterRender,
      contextMenu: true
  });

  hot.render();
```
:::</code></pre>

would be transformed to:
```
const hotRef = React.createRef();
const data = [
  ['test', 'data']
];
const hotSettings = {
  data,
  nestedRows: true,
  afterRender: onAfterRender,
  contextMenu: true
};

useEffect(() => {
  const hot = hotRef.current.hotInstance;

  hot.render();
});

<HotTable ref={hotRef} settings={hotSettings}>
</HotTable>
```
  
### 2. Modifying the existing `::: example` container.
Any code present inside a ` ```js` or ` ```javascript` block wrapped in the `::: examples` container will be picked up and transformed by the transformer.

The final snippet **will** include the imports and the app structure, as that code **is** meant to be executed directly (either by the `Preview` tab or the `Edit` button).

For example, the snippet from point `1` would be transformed into something like:
```
import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import { HotTable } from '@handsontable/react';
import { registerAllModules } from 'handsontable/registry';

// register Handsontable's modules
registerAllModules();

const hotRef = React.createRef();
const data = [
  ['test', 'data']
];
const hotSettings = {
  data,
  nestedRows: true,
  afterRender: onAfterRender,
  contextMenu: true
};

const App = () => {
  useEffect(() => {
    const hot = hotRef.current.hotInstance;

    hot.render();
  });

  return (
    <div>
      <HotTable ref={hotRef} settings={hotSettings}>
      </HotTable>
    </div>
  );
};

ReactDOM.render(<App />, document.getElementById('example'));
```

### Notes:
1. Every transformed snippet will be logged as a `before/after` section in `docs/react-translation-log.html`. The log file is cleared with a new `build` call. This allows easier access to the snippets, as they're scattered between multiple guides and API pages.
2. The transformer does its best to maintain the comments from the original snippets, but it's not always possible. For example, "detached" comments (those not directly above a line of code) are ignored. 
   Single- and multi-line comments (declared by either `//` or `/* ... */`) placed directly above lines of code should be moved to an analogous place in the transformed snippet.
   **Note:** comments placed above Handsontable constructor call will be placed above the declaration of the configuration object in the final snippet.
3. This PR introduces only the React template generator, but the transformer is made the way it could be easily extended to other frameworks.
4. The transformer recognizes the framework by reading an environmental variable by the name of `FRAMEWORK`. 
5. Vuepress utilizes its cache extensively when running the build script, which makes some of the snippets not re-transform on each build. That makes the transformation log file incomplete, as it gets cleared with every new `build` call and changes are logged after snippet transformation. To work around that, I added a `docs:build:no-cache` script, which skips the Vuepress cache.
6. **For the reviewer and tester:** To transform a single code block, skipping the docs building process, you can call the transformation function directly, mocking the `.md`'s file name and line number:
    ```js
    const {
      makeSnippet,
      SnippetTransformer
    } = require('docs/.vuepress/tools/snippet-transform/snippetTransformer');

    const snippetTransformer = new SnippetTransformer('react', snippetContentsAsAString, '', 0);
    const result = snippetTransformer.makeSnippet();
    ```

<sup>[skip changelog]</sup>

### How has this been tested?
Tested the transformer by trying to transform multiple snippets individually and by running `build` and `start` scripts with the `FRAMEWORK` environmental variable set to `react`.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #9312 

### Affected project(s):
- [ ] `handsontable`
- [ ] `@handsontable/angular`
- [ ] `@handsontable/react`
- [ ] `@handsontable/vue`
- [ ] `@handsontable/vue3`

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

wszymanski commented on docs/.vuepress/tools/snippet-transform/snippetTransformer.js:
```diff
@@ -0,0 +1,398 @@
+    parsedContent.body.forEach((node) => {
+      switch (node.type) {
+        case 'ImportDeclaration':
+          // TODO
```
Should it be left?

wszymanski commented on docs/.vuepress/tools/snippet-transform/templates/index.js:
```diff
@@ -0,0 +1,27 @@
+ */
+function renderTemplate(framework, snippetInformation, includeImports = false, includeApp = false, appContainerId) {
+  return (
+    typeof TEMPLATE_RENDERERS[framework] === 'function' ?
```
Extra `if` and returning `null` would be more clear for me. Using parenthesis for me make it unintuitive.

wszymanski reviewed (changes requested):
I've some suggestions . Some of them - related to making code more clear are up to you.

jansiegel commented on docs/.vuepress/tools/snippet-transform/helpers/previewLogger.js:
```diff
@@ -0,0 +1,106 @@
+const path = require('path');
+const fs = require('fs');
+
+const LOG_FILE_NAME = 'react-translation-log.html';
```
Until we make a script to compare the log files, I'll add it to `.gitignore`, but it most probably will be removed from it later on.

warpech replied:
As agreed, this file is not ignored, because the requirements say it must be committed as a way of verifying the correctness of the transformation.

However, the file name is inconsistent. I think it should be `snippet-transformer-log-react.html`

jansiegel commented on docs/.vuepress/helpers.js:
```diff
@@ -69,10 +69,20 @@ function getBuildDocsVersion() {
+ * @returns {string}
+ */
+function getBuildDocsFramework() {
+  return process.env.FRAMEWORK;
```
I'll stick to this for now, as it's easier to test the logic of this PR and deal with the actual helper while merging it to the epic branch, WDYT?

wszymanski replied:
Sure, just please remember that it will need a change.

jansiegel replied:
I forgot about the issue with the env. variable not being available in `docs:start`, so i think I'll retract my previous comment and make a helper for it now.

jansiegel replied:
I copied the entire helper file from #9371 and adapted the code from this PR to it. The helper file will require special attention when merging, but most of the logic is in place.

wszymanski commented on docs/.vuepress/containers/helpers/index.js:
```diff
@@ -0,0 +1,46 @@
+ */
+function getContainerFrontMatterLength(frontMatterObj) {
+  const frontMatterEntriesCount = Object.keys(frontMatterObj).reduce(
+    (sum, key) => {
+      if (Array.isArray(frontMatterObj[key])) {
+        return sum + frontMatterObj[key].length + 1;
+      }
+
+      return sum + 1;
+    }, 0);
```
The same suggestion [as here](https://github.com/handsontable/handsontable/pull/9319#discussion_r859761925).

wszymanski reviewed (commented):
Thanks for your changes. Please sync it with #9307 and #9319 yet and I'll run it again.

jansiegel commented on docs/.vuepress/containers/examples/examples.js:
```diff
@@ -101,7 +109,38 @@ module.exports = {
+
+      // Transform the JS snippet to the framework-based one, where the framework is defined in the `DOCS_FRAMEWORK`
+      // environmental variable.
+      if (!['angular', 'react', 'vue'].some(value => preset.includes(value)) && framework) {
```
Not really, this part is meant to recognize if the container is configured with a framework-based preset, and those don't include `vue3`: https://github.com/handsontable/handsontable/blob/master/docs/README-EDITING.md#adding-code-examples

jansiegel commented:
> Please sync it with #9307 and #9319

I think it'll make more sense to sync it after (or during) the merge. I copied the `helper.js` file to emulate the API available in #9307, but if there are no significant changes in the API naming I'd leave it as is. It'll have to be tested and possibly corrected on the epic branch when the actual build logic will be implemented anyway.

> fix problems found by linter and I'll run it again.

Which problems are you mentioning? It looks to me like the check's clean: https://github.com/handsontable/handsontable/runs/6174207811?check_suite_focus=true.

wszymanski reviewed (commented):
> > Please sync it with #9307 and #9319
> 
> I think it'll make more sense to sync it after (or during) the merge. I copied the `helper.js` file to emulate the API available in #9307, but if there are no significant changes in the API naming I'd leave it as is. It'll have to be tested and possibly corrected on the epic branch when the actual build logic will be implemented anyway.

Ok, merge has been done.

> > fix problems found by linter and I'll run it again.
> 
> Which problems are you mentioning? It looks to me like the check's clean: [handsontable/handsontable/runs/6174207811?check_suite_focus=true](https://github.com/handsontable/handsontable/runs/6174207811?check_suite_focus=true).

I've seen some problems: https://github.com/handsontable/handsontable/runs/6189097864?check_suite_focus=true, https://github.com/handsontable/handsontable/runs/6189097547?check_suite_focus=true, but it seems they are gone for now.

wszymanski reviewed (changes requested):
It seems that there is a problem for `npm run docs:start` script.

```
./.watch-tmp/next/angular-data-grid/guides/internationalization/language.md
Module build failed (from ./node_modules/@vuepress/markdown-loader/index.js):
Error: ENOENT: no such file or directory, open '/Users/aviox/apps/handsontable/docs/react-translation-log.html'
    at Object.openSync (node:fs:585:3)
    at Object.readFileSync (node:fs:453:35)
    at logChange (/Users/aviox/apps/handsontable/docs/.vuepress/tools/snippet-transform/helpers/previewLogger.js:81:22)
    at Object.render (/Users/aviox/apps/handsontable/docs/.vuepress/containers/examples/examples.js:126:9)
    at Renderer.render (/Users/aviox/apps/handsontable/docs/.vuepress/config.js:130:46)
    at Renderer.md.renderer.render (/Users/aviox/apps/handsontable/docs/.vuepress/config.js:140:63)
    at MarkdownIt.render (/Users/aviox/apps/handsontable/docs/node_modules/@vuepress/markdown/node_modules/markdown-it/lib/index.js:543:24)
    at MarkdownIt.md.render (/Users/aviox/apps/handsontable/docs/node_modules/@vuepress/markdown/index.js:142:25)
    at Object.module.exports (/Users/aviox/apps/handsontable/docs/node_modules/@vuepress/markdown-loader/index.js:74:16)
```

wszymanski reviewed (changes requested):
Just small changes suggestion and list of examples which use `hot` and there are some problems:

1. http://localhost:8080/docs/next/react-data-grid/row-prepopulating/ - HOT doesn't work after double click
2. http://localhost:8080/docs/next/react-data-grid/saving-data/#saving-changes-using-a-callback - an error is thrown after after click on the `Save data` button
3. http://localhost:8080/docs/next/react-data-grid/grid-size/#manual-resizing - the demo doesn't work
4. http://localhost:8080/docs/next/react-data-grid/column-filter/#filter-as-you-type - filtering using header input doesn't work
5. http://localhost:8080/docs/next/react-data-grid/cell-renderer/#rendering-custom-html-in-header - the demo doesn't work
6. http://localhost:8080/docs/next/react-data-grid/selection/#selecting-ranges - change in a drop-down throws an error
7. http://localhost:8080/docs/next/react-data-grid/selection/#getting-data-from-the-selected-ranges - click on a button throws an error
8. http://localhost:8080/docs/next/react-data-grid/selection/#modifying-the-selected-cells - as above
9. http://localhost:8080/docs/next/react-data-grid/basic-clipboard/#trigger-copy-cut-programmatically - as above
10. http://localhost:8080/docs/next/react-data-grid/batch-operations/#live-demo-of-the-suspend-feature - as above
11. http://localhost:8080/docs/next/react-data-grid/searching-values/ - filtering don't work
12. http://localhost:8080/docs/next/react-data-grid/export-to-csv/ - click on a button throws an error

jansiegel commented on docs/.vuepress/tools/snippet-transform/snippetTransformer.js:
```diff
@@ -0,0 +1,407 @@
+ * @type {string[]}
+ */
+const SUPPORTED_FRAMEWORKS = [
+  FRAMEWORKS.js,
+  FRAMEWORKS.react
```
I think it's better to store the string values for the framework names in a single constant, rather than use them directly in the code. That way, if we ever decide to use `"js"` instead of `"javascript"`, we will have to change it in a single place, not search for all the occurrences scattered across the code.
Why does this bother you?

wszymanski replied:
I've thought that `FRAMEWORKS.js/FRAMEWORKS.react` is used only to creating `SUPPORTED_FRAMEWORKS` array. It's not. I'm sorry! It's okey then.

warpech commented on docs/.gitignore:
```diff
@@ -12,6 +12,7 @@ config.local.js
 basement_dist
 .idea
 report-check-links.xlsx
+react-translation-log.html
```
According to the project requirements, this file should not be ignored:

> This file will be committed to the repo to allow detecting regressions in the transformation script.

This is especially important since `snippet-transform` does not have a test suite.

jansiegel replied:
I generated the initial translation log and committed in here: 5b558ceb1a9475452e55019cea4f1358938dc73b

jansiegel commented:
> Just small changes suggestion and list of examples which use `hot` and there are some problems:
> 
> 1. http://localhost:8080/docs/next/react-data-grid/row-prepopulating/ - HOT doesn't work after double click
> 2. http://localhost:8080/docs/next/react-data-grid/saving-data/#saving-changes-using-a-callback - an error is thrown after after click on the `Save data` button
> 3. http://localhost:8080/docs/next/react-data-grid/grid-size/#manual-resizing - the demo doesn't work
> 4. http://localhost:8080/docs/next/react-data-grid/column-filter/#filter-as-you-type - filtering using header input doesn't work
> 5. http://localhost:8080/docs/next/react-data-grid/cell-renderer/#rendering-custom-html-in-header - the demo doesn't work
> 6. http://localhost:8080/docs/next/react-data-grid/selection/#selecting-ranges - change in a drop-down throws an error
> 7. http://localhost:8080/docs/next/react-data-grid/selection/#getting-data-from-the-selected-ranges - click on a button throws an error
> 8. http://localhost:8080/docs/next/react-data-grid/selection/#modifying-the-selected-cells - as above
> 9. http://localhost:8080/docs/next/react-data-grid/basic-clipboard/#trigger-copy-cut-programmatically - as above
> 10. http://localhost:8080/docs/next/react-data-grid/batch-operations/#live-demo-of-the-suspend-feature - as above
> 11. http://localhost:8080/docs/next/react-data-grid/searching-values/ - filtering don't work
> 12. http://localhost:8080/docs/next/react-data-grid/export-to-csv/ - click on a button throws an error

The longer I analyze those examples, the more I think they should be reorganized and rewritten for both vanilla and React versions of the docs.

Maybe we could merge this PR as a base for future modifications and correct/rewrite the problematic examples in a separate PR? What do you think? @wszymanski

wszymanski commented:
> The longer I analyze those examples, the more I think they should be reorganized and rewritten for both vanilla and React versions of the docs.

Huh.

> Maybe we could merge this PR as a base for future modifications and correct/rewrite the problematic examples in a separate PR? What do you think? @wszymanski

I'm not sure. Good job is done here. We made merge for not fully working #9307. The `feature/react` branch currently look much better and is probably (?) ready for merge, but I assume that it not possible to release it without examples transpilation. Thus, I would merge, but probably it's question for @warpech? WDYT, @warpech?

warpech commented:
I agree we can merge without improving all the transformations in this PR, but not until all review comments are resolved or tracked elsewhere. For example, my comment https://github.com/handsontable/handsontable/pull/9384#pullrequestreview-969965149 is not resolved.

warpech commented:
Problematic examples found by @wszymanski are now tracked in https://docs.google.com/spreadsheets/d/15q2Bv4UCIoXLktZDDqSTupTLSSO0BWvoxfJbeUbCauU/edit#gid=0

jansiegel commented:
> but not until all review comments are resolved or tracked elsewhere

@warpech @wszymanski I think I addressed all of the comments now? If I'm missing anything, let me know.

warpech commented on docs/react-translation-log.html:
```diff
@@ -0,0 +1,11456 @@
+<!DOCTYPE html>
```
I think that it only makes sense to keep the `react-translation-log.html` file in context of the version that was used as the source, e.g.:

- `docs/next/react-translation-log.html`
- `docs/12.1/react-translation-log.html`
 
Otherwise the files will wrongly overwrite each other.

jansiegel replied:
You're right, but this will require adding some logic to split the logs between versions, when the entire `docs:build` is called.

warpech replied:
Please do it, because otherwise we will get problematic, unmeaningful changes or conflicts in this file. It might be in a separate PR if you prefer.

warpech commented on docs/.vuepress/containers/examples/examples.js:
```diff
@@ -7,6 +7,19 @@ const EXAMPLE_REGEX = /^(example)\s*(#\S*|)\s*(\.\S*|)\s*(:\S*|)\s*([\S|\s]*)$/;
+  SnippetTransformer,
+  logChange,
+  SUPPORTED_FRAMEWORKS
+} = require('../../tools/snippet-transform/snippetTransformer');
```
Wouldn't it be consistent to rename the folder from `snippet-transform` to `snippet-transformer`?

jansiegel replied:
I was trying to be consistent with the other folders in `./.vuepress/tools` - `jsdoc-convert` and `version`, but renaming it to `snippet-transformer` sounds good to me 👍

warpech commented on docs/.vuepress/containers/examples/examples.js:
```diff
@@ -101,7 +114,52 @@ module.exports = {
+        const frontMatterLength = getContainerFrontMatterLength(env.frontmatter);
+        const lineNumber = tokens[jsIndex].map[0] + frontMatterLength;
+        const snippetTransformer = new SnippetTransformer(framework, jsContent, filePath, lineNumber);
+        const translatedSnippetContent = snippetTransformer.makeSnippet(true, true, id);
```
Can it be more consistent: `translated` -> `transformed`

warpech commented on docs/.vuepress/containers/snippets/snippet.js:
```diff
@@ -0,0 +1,64 @@
+        lineNumber
+      );
+
+      const translatedSnippetContent = snippetTransformer.makeSnippet();
```
Can it be more consistent: `translated` -> `transformed`

warpech commented on docs/.vuepress/tools/snippet-transform/helpers/previewLogger.js:
```diff
@@ -0,0 +1,106 @@
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title>Snippet Translation Preview</title>
```
Consistency. Should be: `Snippet Transformer Log`

warpech commented on docs/.vuepress/tools/build.mjs:
```diff
@@ -10,6 +10,7 @@ import {
   getVersions,
   FRAMEWORK_SUFFIX
 } from '../helpers.js';
+import { initLog as initSnippetTransformationLog } from './snippet-transform/helpers/previewLogger.js';
```
I think it would be more consistent if it was called `initSnippetTransformerLog`

warpech commented on docs/.vuepress/tools/snippet-transform/templates/react.js:
```diff
@@ -0,0 +1,150 @@
+// register Handsontable's modules
+registerAllModules();
+
+` : '';
```
Can we please persistently disable cache then? I guess it is even more problematic if you switch branches and end up reusing cache from an old branch.

Cache speeds up `npm run docs:build` and the initialization of `npm run docs:start`, at the expense of being sometimes incorrect. I think it is unacceptable.

@kirszenbaum do you have an opinion here?

kirszenbaum replied:
Interestingly enough, I don't think I ever used the `:no-cache` option, because I didn't have any problems with cache so far. But if `:no-cache` doesn't slow things down a lot, and is theoretically more accurate, I'd go for it

jansiegel replied:
I think the cache kicks in when nothing is modified in the `.md` files, so it probably doesn't matter as much when preparing the actual content.
I'm not sure how much time we save by using the cached script, compared to the non-cached one, though.

jansiegel replied:
@warpech @kirszenbaum 
I did some testing, I consecutively ran:
1. `DOCS_VERSION npm run docs:start` - it took around **4 minutes, 54 seconds**
2. `DOCS_VERSION npm run docs:start` - it took around **54 seconds**
3. `DOCS_VERSION npm run docs:start` - it took around **52 seconds**
4. `DOCS_VERSION npm run docs:start` - it took around **52 seconds**
5. `DOCS_VERSION npm run docs:start:no-cache` - it took around **5 minutes 12 seconds**

So my take on this is - depending on the state of the cache, the watch script can speed up up to ~6 times. I think it's a significant difference.

warpech replied:
OK, let's keep the cache then!

jansiegel commented:
@warpech I think I addressed every new suggestion apart from https://github.com/handsontable/handsontable/pull/9384#discussion_r870756879 in https://github.com/handsontable/handsontable/pull/9384/commits/ffc3101623a15df4082cef77e323f502dc5191d7.

warpech reviewed (approved):
Thanks for the fixes!
